go/src/koding/kites/kloud/provider/softlayer/stack.go

```diff
@@ -0,0 +1,213 @@
+	"bytes"
+	"errors"
+	"fmt"
+	"strconv"
+	"text/template"
+
+	"koding/kites/kloud/stack"
+	"koding/kites/kloud/stack/provider"
+	"koding/kites/kloud/userdata"
+
+	softlayerGo "github.com/maximilien/softlayer-go/client"
+)
+
+//go:generate $GOPATH/bin/go-bindata -mode 420 -modtime 1475345133 -pkg softlayer -o bootstrap.json.tmpl.go bootstrap.json.tmpl
+//go:generate go fmt bootstrap.json.tmpl.go
+
+var (
+	_ provider.Stack = (*Stack)(nil)
+	_ stack.Stacker  = (*Stack)(nil)
+)
+
+func mustAsset(s string) string {
+	a, err := Asset(s)
+	if err != nil {
+		panic(err)
+	}
+	return string(a)
+}
+
+var BootstrapTemplate = template.Must(template.New("").Parse(mustAsset("bootstrap.json.tmpl")))
+
+type bootstrapVars struct {
+	KeyName   string
+	PublicKey string
+}
+
+// Responsible for building/updating/destroying Softlayer's stack
+type Stack struct {
+	*provider.BaseStack
+}
+
+// Verifies the given Softlayer credential.
+// Compared against what is defined in schema.Credential.
+func (s *Stack) VerifyCredential(c *stack.Credential) error {
+	cred, ok := c.Credential.(*Credential)
+	if !ok {
+		return fmt.Errorf("credential is not of type softlayer.Credential: %T", c.Credential)
+	}
+
+	if err := cred.Valid(); err != nil {
+		return err
+	}
+
+	// Do a quick check to verify the credentials are valid for
+	// communitcating with the SoftLayer apis
+	client := softlayerGo.NewSoftLayerClient(cred.Username, cred.ApiKey)
+	service, err := client.GetSoftLayer_Virtual_Guest_Service()
+	if err != nil {
+		return err
+	}
+	_, err = service.GetObject(1234567)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Stack) BootstrapArg() *stack.BootstrapRequest {
+	return s.BaseStack.Arg.(*stack.BootstrapRequest)
+}
+
+// BootstrapTemplates creates the template that is used to generate default
+// output values, to bootstrap a SoftLayer stack
+func (s *Stack) BootstrapTemplates(c *stack.Credential) ([]*stack.Template, error) {
+	vs := &bootstrapVars{
+		KeyName: fmt.Sprintf(
+			"koding-%s",
+			c.Identifier,
+		),
+		PublicKey: s.Keys.PublicKey,
+	}
+
+	var bootstrap bytes.Buffer
+	if err := BootstrapTemplate.Execute(&bootstrap, vs); err != nil {
+		return nil, err
+	}
+
+	// NOTE: Should we shadow out private vars here?
+
+	return []*stack.Template{{
+		Content: bootstrap.String(),
+	}}, nil
+}
+
+// injectBootstrap injects bootstrap resources into a SoftLayer stack, which
+// is to say, it provides specific default values if user doesn't provide them
+func (s *Stack) injectBootstrap(b *Bootstrap, guest map[string]interface{}) error {
+	bootstrapKey, err := strconv.Atoi(b.KeyID)
+	if err != nil {
+		return err
+	}
+
+	// Apply user ssh keys if provided, otherwise use the bootstrap key
+	keys := []int{bootstrapKey}
+
+	if userKeys, ok := guest["ssh_keys"].([]int); ok {
+		keys = append(keys, userKeys...)
+	}
+	guest["ssh_keys"] = keys
+
+	guest["image"] = "UBUNTU_14_64"
+	// NOTE: Should we allow for user override here? Do all images work with klient?
```

Comment:
Bootstrap templates are fully controlled by kloud currently (server-side), user can't edit it now.

We're working actively to make klient support more distros, just FYI.